[WIP] Add comments/documentation on Network Fabric

matte21 opened the pull request:
Some information is still missing:
- What happens on concurrent duplicate/conflicting deletes
- Some information on how the OvS fabric does the listing of existing local/remote interfaces

matte21 committed:
Add comments/documentation on Network Fabric

MikeSpreitzer commented on staging/kos/pkg/networkfabric/interfaces.go:
```diff
@@ -18,33 +18,100 @@ package networkfabric
+// Network Interfaces are identified by a name, which must be unique over space.
+// (VNI, guest IP) pairs must also be unique over space, that is, for this
+// contract two Network Interfaces with the same (VNI, guest IP) pair are the
+// same Network Interface. Follow some guarantees that implementers MUST make:
```
What about (VNI, GuestMAC)?

matte21 replied:
On the basis that a VNI identifies a VXLAN segment, and a VXLAN segment represents a Layer2 network, where IDs are MAC addresses, as opposed to IP addresses?

MikeSpreitzer replied:
yes

matte21 replied:
Sounds good, will change it tomorrow.

MikeSpreitzer commented:
What happens if the process running this code crashes in the midst of making a change?
What happens if there are two processes running this code at the same time?

matte21 committed:
Add missing information requested in review

Also, minor clean ups.

matte21 commented:
c090a0a adds the requested information.

> What happens if the process running this code crashes in the midst of making a change?

An incomplete Network Interface implementation might be left on the node. Precisely, that can happen only for local interfaces; the incomplete implementation consists of the Linux network device w/o the OpenFlow flows. `ListLocalIfcs()` detects such incomplete implementations and cleans them up by deleting the Linux network device, so any fabric user that invokes `ListLocalIfcs()` before doing anything else with the fabric is safe.  

> What happens if there are two processes running this code at the same time?

The fabric is not robust to this scenario, I came up with two ways in which things can go wrong and break things very badly, and I didn't think about this much, so I expect there are more. I'd rather add the safety check after the perf study. Any idea on how to prevent this?
I thought of lock files through https://github.com/juju/fslock, but apparently lock files in general are unreliable on Linux: https://www.kernel.org/doc/Documentation/filesystems/mandatory-locking.txt . Maybe making the fabric open a Unix Socket only for this purpose (I know, it's ugly)?

MikeSpreitzer commented on staging/kos/pkg/networkfabric/interfaces.go:
```diff
@@ -18,33 +18,116 @@ package networkfabric
-// the user to implement Netowrk Interfaces. The implementer
-// MUST return nil when the user attempts to delete an Interface
-// which does not exist.
+// Interface is the contract of a VXLAN network fabric.
```
"Interface" is a very reasonable name in general, but in this context is confusing because we also have the prior term "network interface".

matte21 replied:
How about "Contract"?

matte21 replied:
switched to "Contract" in bfbf60c

MikeSpreitzer replied:
What about "Manager" or "Factory" or "InterfaceFactory"?

matte21 replied:
"Manager" seems good enough to me. "Factory"  would be confusing, as we also have a NetworkFabric factory.

matte21 replied:
Actually, I think "Contract" is better than "Manager". The package name is "networkfabric". For importers explicitly using the network fabric interface (as in, programming language interface), "networkfabric.Contract" is less misleading than "networkfabric.Manager". I think the latter might be interpreted as "a manager of networkfabrics", which is not, it's a manager of network interfaces.

We could rename the package to "networkinterface" and "Contract" to "Manager". This way, importers would use "networkinterface.Manager", that conveys the right idea.

MikeSpreitzer replied:
I like the package name networkfabric.  There is more to the fabric than the network interfaces.

What about `networkfabric.InterfaceManager`?

matte21 replied:
Ok.

matte21 replied:
92694a5 does the update.

MikeSpreitzer commented on staging/kos/pkg/networkfabric/interfaces.go:
```diff
@@ -18,33 +18,116 @@ package networkfabric
+//
+// A local Network Interface is the networking state of a guest that is bound to
+// the same node as the user of this contract. Creating a local Network
+// Interface means creating a Linux network device and configuring the node's
```
I think the more familiar term in Linux is "network interface".  "device" suggests something found in `/dev`.

matte21 replied:
e919b51 performs the suggested rewording. I do wonder if the new wording is still confusing though. Throughout the comments (and in function names) the hard state created by the fabric is called "local Network Interface". This leads to comments such as "a local Network Interface is implemented as a Linux network interface...".

MikeSpreitzer replied:
Yeah, the problem is that we need a technical term for our software artifacts that is not the same as "network interface" because one of our things is more than a network interface (it includes the flows) and, in some (remote) cases, less.  If we had one golang type, we could use that as the term.  But we have two (LocalNetIfc and RemoteNetIfc).

matte21 replied:
I think it's ok to use the names of the Golang types, because the comments always make a distinction between "local Network Interface" and "remote Network Interface", hence there's no need for a catch-all name.

matte21 committed:
s/Linux network device/Linux network interface/

matte21 committed:
Rename netfabric.Interface to netfabric.Contract

Rename netfabric.Interface to netfabric.Contract to avoid confusion,
because the netfabric.Interface deals with ...network interfaces.

Also, do some minor renamings.

MikeSpreitzer commented on staging/kos/pkg/networkfabric/ovs/doc.go:
```diff
@@ -0,0 +1,94 @@
+// through OpenFlow.
+//
+//******************************************************************************
+// Currently, THE FABRIC DOES NOT WORK IF ANOTHER OvS BRIDGE EXISTS ON THE NODE.
```
This is surprisingly fragile.  Couldn't we remove this restriction by using a bridge with a fixed name?

matte21 replied:
The bridge already has a fixed name. The problem is, when listing existing local network interfaces, you need to query all the existing interfaces connected to that bridge, and IIRC when I wrote that code the OvS CLI did not have a straightforward way to do this. So what is done in the code is listing all the network interfaces managed by the OvS daemon, that is, all the network interfaces of all the OvS bridges on the node. I'm sure there is a way to get the behavior we want, but the OvS CLI doesn't make our life easy in that regard. Will dig more into this ASAP (probably later today).

matte21 replied:
Actually, I did some research and I think there's a reasonably simple way to remove the fragility. Will get to it later today.

matte21 committed:
Rename: netfabric.Contract -> netfabric.InterfaceManager
